### PR TITLE
x64 windows sourcehook hookmangen

### PR DESCRIPTION
**Pull request by Kenzzer**
Add x64 dynamic hook support to SourceHook, Windows only

- Added `x64GenContext` to generate SourceHook virtual stubs on x86_64
- `GenBuffer` has been moved to `sh_asm.h`, since it's only ever used to allocate blocks of memories where assembly is written at
- `x64JitWriter` (inheriting from `GenBuffer`). It's a collection of helper functions to write x86_64 assembly bytes
- Added `IGenContext` in order to keep a single implementation of `CHookManagerAutoGen` that can be compiled on all architectures. Consequently, `x64GenContext` & `GenContext` inherit from `IGenContext`

This is already a **massive** PR as it is. And I really don't want to be adding Linux support right now, but it will come shortly after rest assured.

### CODE REVIEW

**Comment by Kenzzer** [comment]
@dvander @TheDS @peace-maker 

I believe you'll be the ones that want to look most at this, given it's pertaining to sourcehook. My goal is to have this PR merged, before I make any further changes for linux support.

File: core/metamod.cpp [hunk 1]
```diff
@@ -417,11 +417,17 @@ FileSystemFactory(const char *iface, int *ret)
 void
 mm_LogMessage(const char *msg, ...)
 {
+	static bool g_logging = false;
+	if (g_logging) {
+		return;
+	}
+	g_logging = true;
```
Comment (Kenzzer) [inline core/metamod.cpp:424]: This change exist, because I happened to be log looping to hell while debugging sourcehook.

File: core/metamod.cpp [hunk 2]
```diff
@@ -124,7 +124,7 @@ static ConVar *mm_basedir = NULL;
 static CreateInterfaceFn engine_factory = NULL;
 static CreateInterfaceFn physics_factory = NULL;
 static CreateInterfaceFn filesystem_factory = NULL;
-#if !defined( _WIN64 ) && !defined( __amd64__ )
+#if !defined( PLATFORM_64BITS ) && !defined( _LINUX )
```
Comment (dvander) [inline core/metamod.cpp:127]: Using platform macros is preferred over Valve-isms.

File: core/metamod.cpp (hunk 1 above)
Comment (dvander) [reply core/metamod.cpp:424]: Use RAII. (I think MM:S doesn't use AMTL, but if it does, there's a helper for this pattern called SaveAndSet).

File: core/metamod.cpp [hunk 3]
```diff
@@ -417,11 +417,17 @@ FileSystemFactory(const char *iface, int *ret)
...
+	if (g_logging) {
+		return;
+	}
+	g_logging = true;
 	va_list ap;
 	static char buffer[2048];
 
 	va_start(ap, msg);
 	size_t len = vsnprintf(buffer, sizeof(buffer) - 2, msg, ap);
+	len = std::min<size_t>(len, 2046);
```
Comment (dvander) [inline core/metamod.cpp:430]: Where does this magic number come from?

**Review by dvander — APPROVED** [review]
I only took a brief look at this, but it looks good on the whole. Some small suggestions inline.

File: core/metamod.cpp (hunk 3 above)
Comment (Kenzzer) [reply core/metamod.cpp:430]: Just above, `sizeof(buffer) - 2` though regardless I shouldn't have hard programmed it.

Issue is with the return value of `vsnprintf`, just below we do this.
```
buffer[len++] = '\n';
buffer[len] = '\0';
```
Which fine so as long as `vsnprintf` never returns a number above 2046. However while it's true `vsnprintf` returns the amount of bytes written, it also has another scenario as per c++ documentation : 
> Return Value :
> 4) Number of characters written if successful or negative value if an error occurred. If the resulting string gets truncated due to buf_size limit, function returns the total number of characters (not including the terminating null-byte) which would have been written, if the limit was not imposed.

So if it returns 2047 or above, that's an instant crash when we do
```
buffer[len++] = '\n';
buffer[len] = '\0';
```

Anyways gonna change the value to `sizeof(buffer) - 2`, so there's no magic number.

**Pull request merged**